[GTK][WPE] OpenGL assets created by WebGL are not released when destroying the WebGLRenderingContext

magomez opened the pull request:
#### 1d16639f9d572ad6d75224c49b9cf0ca598e3bbc
<pre>
[GTK][WPE] OpenGL assets created by WebGL are not released when destroying the WebGLRenderingContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=300085">https://bugs.webkit.org/show_bug.cgi?id=300085</a>

Reviewed by Carlos Garcia Campos.

Stop using a singleton for the ANGLE sharing context. Each ANGLE context will use
its own ANGLE sharing context instead, that will be created and destroyed together
with the ANGLE context. This ensures that all the OpenGL resources that were created
by ANGLE will be destroyed when the contexts are destroyed.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::clearGLContexts):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/angle/PlatformDisplayANGLE.cpp:
(WebCore::PlatformDisplay::angleSharingGLContext):
(WebCore::PlatformDisplay::clearANGLESharingGLContext): Deleted.
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLTextureMapperANGLE::platformInitializeContext):

Canonical link: <a href="https://commits.webkit.org/301043@main">https://commits.webkit.org/301043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94f3132e980d37be7a2d21afde0b31aea997f23d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124378 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34788 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131217 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126255 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44804 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94614 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127332 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35709 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111243 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75200 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34645 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29403 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74694 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105454 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29625 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133883 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39108 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103094 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51675 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107461 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102889 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48252 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26501 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48207 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19562 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51143 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56928 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50570 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53929 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52244 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
```diff
@@ -107,10 +107,16 @@ GraphicsContextGLANGLE::~GraphicsContextGLANGLE()
             ASSERT_UNUSED(result, !!result);
         }
-        EGL_MakeCurrent(m_displayObj, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
         EGL_DestroyContext(m_displayObj, m_contextObj);
```
Don't we need to keep that to make sure the context is not current when destroying it?

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
```diff
@@ -107,10 +107,16 @@ GraphicsContextGLANGLE::~GraphicsContextGLANGLE()
+    if (m_angleSharingContextObj)
+        EGL_DestroyContext(m_displayObj, m_angleSharingContextObj);
+
+    // Ideally this should go before the m_contextObj destruction, but there are platforms where it breaks
+    // the destruction m_contextObj. Putting it here works for all the platforms that I've tested.
+    EGL_MakeCurrent(m_displayObj, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
```
Ah, I see, this is weird, because it worked before, no?

magomez replied:
It works on most of the platforms, but it doesn't on some platforms with different drivers.

The curious thing is that the first thing EGL_DestroyContext does is making the context to be destroyed current, in order to release the resources allocated by it. This is what fails on some platforms. Removing the current context just to call EGL_DestroyContext and make it current again seems to be a problem.

In any case, if we don't make the context not current and just call EGL_DestroyContext, the context is marked for destruction, and is really destroyed as soon as it stops being current.

magomez committed:
[GTK][WPE] OpenGL assets created by WebGL are not released when destroying the WebGLRenderingContext
https://bugs.webkit.org/show_bug.cgi?id=300085

Reviewed by Carlos Garcia Campos.

Stop using a singleton for the ANGLE sharing context. Each ANGLE context will use
its own ANGLE sharing context instead, that will be created and destroyed together
with the ANGLE context. This ensures that all the OpenGL resources that were created
by ANGLE will be destroyed when the contexts are destroyed.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::clearGLContexts):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/angle/PlatformDisplayANGLE.cpp:
(WebCore::PlatformDisplay::angleSharingGLContext):
(WebCore::PlatformDisplay::clearANGLESharingGLContext): Deleted.
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLTextureMapperANGLE::platformInitializeContext):

Canonical link: https://commits.webkit.org/301043@main
